nexus/src/app/instance.rs

```diff
@@ -1132,4 +1140,202 @@ impl super::Nexus {
+                Some(msg) => {
+                    (Fuse::terminated(), propolis_sink.send(msg).fuse())
+                }
+            };
+            let (nexus_write, propolis_read) = match buffered_output.take() {
+                None => (Fuse::terminated(), propolis_stream.next().fuse()),
+                Some(msg) => (nexus_sink.send(msg).fuse(), Fuse::terminated()),
+            };
+            tokio::select! {
+                msg = nexus_read => {
+                    match msg {
+                        None => break,
+                        Some(Err(e)) => return Err(e)?,
+                        Some(Ok(WebSocketMessage::Close(details))) => {
+                            buffered_input = Some(WebSocketMessage::Close(details));
+                            break
+                        }
+                        Some(Ok(WebSocketMessage::Text(_text))) => {
+                            // TODO: json payloads specifying client-sent metadata?
+                        }
+                        Some(Ok(WebSocketMessage::Binary(data))) => {
+                            buffered_input = Some(WebSocketMessage::Binary(data))
+                        }
+                        // Frame won't exist at this level, and ping reply is handled by tungstenite
+                        Some(Ok(WebSocketMessage::Frame(_) | WebSocketMessage::Ping(_) | WebSocketMessage::Pong(_))) => {}
+                    }
+                }
+                result = nexus_write => {
+                    result?;
+                }
+                msg = propolis_read => {
+                    match msg {
+                        None => break,
+                        Some(Err(e)) => return Err(e)?,
+                        Some(Ok(WebSocketMessage::Close(details))) => {
+                            buffered_output = Some(WebSocketMessage::Close(details));
+                            break
+                        }
+                        Some(Ok(WebSocketMessage::Text(_text))) => {
+                            // TODO: deserialize a json payload, specifying:
+                            //  - event: "migration"
+                            //  - address: the address of the new propolis-server
+                            //  - offset: what byte offset to start from (the last one sent from old propolis)
+                        }
+                        Some(Ok(WebSocketMessage::Binary(data))) => {
+                            buffered_output = Some(WebSocketMessage::Binary(data))
+                        }
+                        // Frame won't exist at this level, and ping reply is handled by tungstenite
+                        Some(Ok(WebSocketMessage::Frame(_) | WebSocketMessage::Ping(_) | WebSocketMessage::Pong(_))) => {}
+                    }
+                }
+                result = propolis_write => {
+                    result?;
+                }
+            }
+        }
+        // pass along Close messages
```

Comment:
Should we also try to gracefully send a close to the other side if one unexpectedly drops?